core/src/main/java/org/apache/iceberg/InheritableMetadataFactory.java

```diff
@@ -30,22 +31,40 @@ static InheritableMetadata empty() {
   }
 
   static InheritableMetadata fromManifest(ManifestFile manifest) {
-    return new BaseInheritableMetadata(manifest.snapshotId());
+    if (manifest.snapshotId() != null) {
+      return new BaseInheritableMetadata(manifest.snapshotId(), manifest.sequenceNumber());
+    } else {
+      return NOOP;
```

Comment:
I'm not sure in which case we will encounter the snapshot == null, mind to explain the case ? Thanks.

Reply:
Snapshot id is null when a manifest is appended to a table and the snapshot id hasn't been assigned yet.

If the manifest is committed to table metadata, then it will be set when writing and will always be present. That's why it is required in the v2 schema for manifest list files.

Some appended manifests are rewritten before committing. When reading those to rewrite them, this path is used.

I think I'm going to update how this works, but as a separate commit. The problem is that this allows reading a manifest without filling in the snapshot id. But the rewrite method where the reader that is configured this was is used has the snapshot id. So we should add the snapshot id to the `ManifestFile` and then read to avoid the special case.

Reply:
`added_snapshot_id` is optional because we used to store manifests as a list in the metadata.json file. V2 changes it to required because we always write a manifest-list with the inherited snapshot id and sequence number.

My comment about changing this wasn't about removing `InheritableMetadata`. It was about the path that rewrites the manifest and uses the no-op `InheritableMetadata`. I think that we can simplify this by making sure we set the snapshot ID on the `ManifestFile` before rewriting. That way any rewrite has the snapshot ID and we don't need a special case here.

Reply:
I asked about `InheritableMetadata` as I wondered myself whether we can remove that completely. Seems not. 

Yeah, I like the idea of setting the snapshot id in `ManifestFile` before rewriting.

Reply:
I'll look into removing it entirely, although I don't mind it very much. As long as it is baked into the reader and unavoidable, it seems like a clean way to apply the inherited values.